quinn-proto/src/endpoint.rs

```diff
@@ -188,17 +195,52 @@ impl Endpoint {
         //
 
         let addresses = FourTuple { remote, local_ip };
-        if let Some(ch) = self.index.get(&addresses, &first_decode) {
-            return Some(DatagramEvent::ConnectionEvent(
-                ch,
-                ConnectionEvent(ConnectionEventInner::Datagram {
-                    now,
-                    remote: addresses.remote,
-                    ecn,
-                    first_decode,
-                    remaining,
-                }),
-            ));
+        if let Some(route_to) = self.index.get(&addresses, &first_decode) {
+            let event = DatagramConnectionEvent {
+                now,
+                remote: addresses.remote,
+                ecn,
+                first_decode,
+                remaining,
+            };
+            match route_to {
+                RouteDatagramTo::Incoming(incoming_idx) => {
+                    let incoming_buffer = &mut self.incoming_buffers[incoming_idx];
+                    let config = &self.server_config.as_ref().unwrap().transport;
+                    let buffer_limit = config.receive_window.into_inner().min(
+                        config
+                            .max_concurrent_bidi_streams
+                            .into_inner()
+                            .saturating_add(config.max_concurrent_uni_streams.into_inner())
+                            .saturating_mul(config.stream_receive_window.into_inner()),
+                    );
+
+                    // By checking the buffer limit before rather than after adding the current
+                    // datagram size, we naturally allow for around one datagram worth of QUIC
+                    // overhead before we start dropping packets.
```

Comment:
Rounding up to the next packet isn't necessarily about allowing for overhead (which might vary dramatically according to sender discretion), just erring on the side of being permissive.

Reply:
Should be addressed, see comment below.